Automated signature-to-apidoc replacement (first batch)

ignasi35 opened the pull request:
References #28901 

when moving from `@reference` to `@apidoc` directives there's a bit of a mechanical work such as locating the correct type, the signatures for java and scala, etc...

I put together a replacement tool so the rewrite is simpler (edit/remove suggestions).

ignasi35 commented:
This is DRAFT but I'm pushing a first result of the replacement tool to get feedback on some results.

ignasi35 commented on akka-docs/src/main/paradox/stream/operators/Flow/asFlowWithContext.md:
```diff
@@ -8,12 +8,11 @@ Turns a Flow into a FlowWithContext which can propagate a context per element al
 
 ## Signature
 
-@@signature [Flow.scala](/akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala) { #asFlowWithContext }
-
+@apidoc[asFlowWithContext](Flow) { scala="#asFlowWithContext[U,CtxU,CtxOut](collapseContext:(U,CtxU)=&gt;In)(extractContext:Out=&gt;CtxOut):akka.stream.scaladsl.FlowWithContext[U,CtxU,Out,CtxOut,Mat]" java="#asFlowWithContext(akka.japi.function.Function2,akka.japi.function.Function)" }
 @@@
```
This is the ideal case for the replacement tool: the `asFlowWithContext` operator only has one implementation, the implementation is on the class `Flow` (not a companion object) and the operator has no overloads.

ignasi35 commented on akka-docs/src/main/paradox/stream/operators/Flow/completionStageFlow.md:
```diff
@@ -8,8 +8,8 @@ Streams the elements through the given future flow once it successfully complete
 
-@@signature [Flow.scala](/akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala) { #futureFlow }
-
+@apidoc[completionStageFlow](Flow$) { java="#completionStageFlow(java.util.concurrent.CompletionStage)" }
+@apidoc[completionStageFlow](Flow) { java="#completionStageFlow(java.util.concurrent.CompletionStage)" }
 @@@
```
I want the auto replacement to remove the `@@@div { .group-scala }`/`@@@` too. I'll work on that soon.

Bonus points: the `@@@div { .group-scala }`/`@@@` filter is wrong on this particular oeprator since it's a java-specific operator.

ignasi35 commented on akka-docs/src/main/paradox/stream/operators/ActorFlow/ask.md:
```diff
@@ -18,8 +18,12 @@ This operator is included in:
+@apidoc[ask](Source) { scala="#ask[S](parallelism:Int)(ref:akka.actor.ActorRef)(implicittimeout:akka.util.Timeout,implicittag:scala.reflect.ClassTag[S]):FlowOps.this.Repr[S]" scala="#ask[S](ref:akka.actor.ActorRef)(implicittimeout:akka.util.Timeout,implicittag:scala.reflect.ClassTag[S]):FlowOps.this.Repr[S]" java="#ask(akka.actor.ActorRef,java.lang.Class,akka.util.Timeout)" java="#ask(int,akka.ac ...
+@apidoc[ask](SubFlow) { scala="#ask[S](parallelism:Int)(ref:akka.actor.ActorRef)(implicittimeout:akka.util.Timeout,implicittag:scala.reflect.ClassTag[S]):FlowOps.this.Repr[S]" scala="#ask[S](ref:akka.actor.ActorRef)(implicittimeout:akka.util.Timeout,implicittag:scala.reflect.ClassTag[S]):FlowOps.this.Repr[S]" }
+@apidoc[ask](Flow) { scala="#ask[S](parallelism:Int)(ref:akka.actor.ActorRef)(implicittimeout:akka.util.Timeout,implicittag:scala.reflect.ClassTag[S]):FlowOps.this.Repr[S]" scala="#ask[S](ref:akka.actor.ActorRef)(implicittimeout:akka.util.Timeout,implicittag:scala.reflect.ClassTag[S]):FlowOps.this.Repr[S]" java="#ask(akka.actor.ActorRef,java.lang.Class,akka.util.Timeout)" java="#ask(int,akka.acto ...
 @@@
```
I am considering a manual trim of the generated output so the more complex cases (such as the `ask` operator above) need to be migrated manually.

ignasi35 commented on akka-docs/src/main/paradox/stream/operators/ActorSink/actorRef.md:
```diff
@@ -41,4 +43,4 @@ limiting operator in front of this `Sink`.
 ## Examples
 
-TODO (in progress)
+TODO (in progress)
```
Hmm, I think the tool remove the trailing `\n`

ignasi35 committed:
Some cleanup to prepare...

ignasi35 commented:
The script I had produced some invalid markdown or some content that caused timeouts on the `paradox` task in `sbt`. I've dropped a few Akka stream operators from this PR and also manually review those I'm submitting here.

ennru commented on akka-docs/src/main/paradox/stream/operators/FileIO/fromFile.md:
```diff
@@ -10,13 +10,9 @@ The `fromFile` operator has been deprecated, use @ref:[fromPath](./fromPath.md)
-@@signature [FileIO.scala](/akka-stream/src/main/scala/akka/stream/scaladsl/FileIO.scala) { #fromFile }
-
-@@@
+@apidoc[fromFile](FileIO$) { scala="#fromFile(f:java.io.File,chunkSize:Int):akka.stream.scaladsl.Source[akka.util.ByteString,scala.concurrent.Future[akka.stream.IOResult]]" java="#fromFile(java.io.File)" }
```
It would make sense to show the class name in the link text.
`@apidoc[FileIO.fromFile](FileIO$)`

ignasi35 replied:
Yeah, good idea.

ennru reviewed (commented):
Great to get these done all at once!

ignasi35 committed:
The automatic replacement produces Markdown that caused the paradox task in sbt to timeout or event fail. So
this commit removes some operators with outpu too complex:

- ignored operators:
          "ask.md",
          "alsoTo.md",
          "batchWeighted.md",
          "buffer.md",
          "actorRef.md",
          "collect.md",
          "collection.md",
          "combine.md",
          "completionTimeout.md",
          "concat.md",
          "from.md",
          "fromMaterializer.md",
          "map.md",
          "merge.md",
          "queue.md",
          "log.md", // too many overloads, breaks `paradox` task
          "throttle.md", // too many overloads, breaks `paradox` task
          "idleTimeout.md", // too many overloads, breaks `paradox` task
          "setup.md",
          "watch.md",
          "withBackoff.md",
          "zip.md",
          "zipWith.md",
          "actorRefWithBackpressure.md"

- Ignored class when FQCN contains: Implicits, FlowOpsMat, SubSource, FlowOps, SubFlow, WithContext, DelayStrategy

- some extra manual cleanup on the committed operators.

ignasi35 commented:
Note that I've aggressively removed the `@@@div { .group-scala }` around `### Signature` since `@apidoc` already handles that wuite well. 

The current solution can lead to a `Signture` section that's completely blank (e.g. `fromFuture` only exists for scala) but I think the change in this PR is good enough (and we can manually fix these particular cases later).

raboof reviewed (approved):
Nice! Tried a number and didn't find any dead links.

Perhaps nice to also share the script in `/scripts/` for future reference?

ennru reviewed (approved):
LGTM.
Needs a rebase.

patriknw commented on akka-docs/src/main/paradox/stream/operators/Flow/asFlowWithContext.md:
```diff
@@ -4,13 +4,10 @@ Turns a Flow into a FlowWithContext which can propagate a context per element al
 ## Signature
 
-@@signature [Flow.scala](/akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala) { #asFlowWithContext }
+@apidoc[Flow.asFlowWithContext](Flow) { scala="#asFlowWithContext[U,CtxU,CtxOut](collapseContext:(U,CtxU)=&gt;In)(extractContext:Out=&gt;CtxOut):akka.stream.scaladsl.FlowWithContext[U,CtxU,Out,CtxOut,Mat]" java="#asFlowWithContext(akka.japi.function.Function2,akka.japi.function.Function)" }
```
glad that we don't have to figure out these signatures manually

patriknw commented:
> Perhaps nice to also share the script in /scripts/ for future reference?

I think that is a requirement. There is no way we can ask contributors (or ourselves) to figure out those signatures manually.

ignasi35 commented:
> I think that is a requirement. There is no way we can ask contributors (or ourselves) to figure out those signatures manually.

The code for this is already available at https://github.com/ignasi35/akka-doc-rewriter/blob/master/src/main/scala/com/example/DocRewriter.scala. It's the opposite of maintainable and nice and requires manually setting up a local copy of scaladocs and apidocs.

The moving away from `signature` is a one shot process and the way to figure out the methods signature is actually simple, just tedious: google the apidoc/scaladoc and copy/paste the anchro link to the method.

The reason why I wanted to automate this was that @ennru always added this change when I sent a PR with some code examples.

There's only 36 pages in the operators section that still use the `signature` directive and I was planning on doing them myself anyway since they cases not easily automated (`ask`, actorRef`,...) because there may be multiple overloads or we need to carefully pick the correct link.

ignasi35 commented:
> > Perhaps nice to also share the script in /scripts/ for future reference?
> 
> I think that is a requirement. There is no way we can ask contributors (or ourselves) to figure out those signatures manually.

Put another way: once I'm done, we won't need to support the `signature` directive anymore.

ignasi35 committed:
Merge branch 'master' into unify-operator-signature-apidoc

patriknw commented:
ok, I think I misunderstood that it was based on the existing `signature` directive, but of course it must take the information from something.

I was thinking that it will be difficult to add these apidoc definitions when new operators are added, but yeah, I guess we'll figure out how to.

patriknw reviewed (approved):
LGTM
